### PR TITLE
feat: [PIPE-22045]: repo file editor in gitness

### PR DESCRIPTION
**Pull request by shaurya-harness**
1. Adds React query caching to file explorer
2. Adds Diff view to file edit
3. Adds Git Commit modal cases : CREATE / MOVE / UPDATE / DELETE with violations check
4. Adds functionality for new file in repo/code pages

Screenshot 2024-10-18 at 7 12 21 PM
Screenshot 2024-10-18 at 7 12 28 PM
Screenshot 2024-10-18 at 7 13 12 PM

### CODE REVIEW

File: apps/gitness/src/components/FileEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+  FILE_SEPERATOR,
+  PLAIN_TEXT
+} from '../utils/git-utils'
+import { PathParams } from '../RouteDefinitions'
+import { PathParts, splitPathWithParents } from '../utils/path-utils'
+import { CodeDiffEditor, CodeEditor } from '@harnessio/yaml-editor'
+import { themes } from '../pages/pipeline-edit/theme/monaco-theme'
+import GitCommitDialog from './GitCommitDialog'
+
+export type ViewTypeValue = 'contents' | 'changes'
```
Comment (cjlee01) [inline apps/gitness/src/components/FileEditor.tsx:34]: can we export this to a types file ?

File: apps/gitness/src/components/FileContentViewer.tsx [hunk 2]
```diff
@@ -124,6 +124,23 @@ export default function FileContentViewer({ repoContent }: FileContentViewerProp
     )
   }
 
+  const CodeView = () => {
```
Comment (cjlee01) [inline apps/gitness/src/components/FileContentViewer.tsx:127]: quick question: how does the code editor handle binaries or too large files?

File: apps/gitness/src/components/FileEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    () => (isNew ? GitCommitAction.CREATE : isUpdate ? GitCommitAction.UPDATE : GitCommitAction.MOVE),
+    [isNew, isUpdate]
+  )
+  const inputRef = useRef<HTMLInputElement | null>()
+
+  useEffect(() => {
+    if (inputRef.current && fileName?.length) {
+      inputRef.current.size = Math.min(Math.max(fileName?.length - 2, 20), 50)
+    }
+  }, [fileName, inputRef])
```
Comment (cjlee01) [inline apps/gitness/src/components/FileEditor.tsx:85]: if u are editing the filename, does this cause a rerender?

File: apps/gitness/src/components/FileEditor.tsx (hunk 1 above)
Comment (shaurya-harness) [reply apps/gitness/src/components/FileEditor.tsx:34]: sure, can we do this in seperate PR ?

File: apps/gitness/src/components/FileEditor.tsx (hunk 3 above)
Comment (shaurya-harness) [reply apps/gitness/src/components/FileEditor.tsx:85]: yes it does

File: apps/gitness/src/components/FileContentViewer.tsx (hunk 2 above)
Comment (shaurya-harness) [reply apps/gitness/src/components/FileContentViewer.tsx:127]: binaries are similar to what you see on github / vs-code, attached a screenshot 
will need to test for long files
@srdjan-harness did you check the code editor for long files?
Screenshot 2024-10-21 at 12 20 20 AM

File: apps/gitness/src/components/FileEditor.tsx (hunk 1 above)
Comment (cjlee01) [reply apps/gitness/src/components/FileEditor.tsx:34]: sure

**Review by cjlee01 — APPROVED** [review]

**Comment by cjlee01** [comment]
for the create a new branch and start a pull request , can you navigate to this page 
```{
                      path: 'compare',
                      element: <CreatePullRequest />
                    }```

**Pull request merged**